File: leetcode/Stack/leetcode-0084/84.go
package stack

import "container/list"

func largestRectangleArea(heights []int) int {
	if heights == nil || len(heights) == 0 {
		return 0
	}
	heights = append(heights, 0)
	i, res := 0, 0
	stack := list.New()
	for i < len(heights) {
		if stack.Len() == 0 || heights[stack.Back().Value.(int)] <= heights[i] {
			stack.PushBack(i)
			i++
		} else {
			tmp := stack.Back().Value.(int)
			stack.Remove(stack.Back())
			local := 0
			if stack.Len() == 0 {
				local = heights[tmp] * i
			} else {
				local = heights[tmp] * (i - stack.Back().Value.(int) - 1)
			}
			if local > res {
				res = local
			}
		}
	}
	return res
}

File: leetcode/Stack/leetcode-0084/84.java
import java.util.Arrays;

class Solution {
    public int largestRectangleArea(int[] heights) {
        if (heights == null || heights.length == 0) {
            return 0;
        }

        int i = 0;
        int res = 0;
        Stack<Integer> stack = new Stack<>();
        int[] h = new int[heights.length + 1];
        h = Arrays.copyOf(heights, heights.length + 1);
        while (i < h.length) {
            if (stack.isEmpty() || h[stack.peek()] <= h[i]) {
                stack.push(i++);
            } else {
                int tmp = stack.pop();
                res = Math.max(res, h[tmp] * (stack.isEmpty() ? i : i - stack.peek() - 1));
            }
        }
        return res;
    }
}